src/gtest/test_coins.cpp

```diff
@@ -831,3 +977,492 @@ TEST(CoinsTests, AnchorsTest)
     anchorsTestImpl<OrchardMerkleFrontier>(ORCHARD);
     }
 }
+
+enum SubtreeAction {
+    FlushCache1,
+    FlushCache2,
+    Pop,
+
+    // Push the nth index subtree to cache2
+    Push0,
+    Push1,
+    Push2,
+    Push3,
+
+    // The latest subtree in cache N equals the index K subtree
+    LatestC1EqualsX, // no value
+    LatestC1Equals0,
+    LatestC1Equals1,
+    LatestC1Equals2,
+    LatestC1Equals3,
+
+    LatestC2EqualsX, // no value
+    LatestC2Equals0,
+    LatestC2Equals1,
+    LatestC2Equals2,
+    LatestC2Equals3,
+
+    LatestBaseEqualsX,
+    LatestBaseEquals0,
+    LatestBaseEquals1,
+    LatestBaseEquals2,
+    LatestBaseEquals3,
+
+    // Get K index from cache N
+    Get0C1,
+    Get1C1,
+    Get2C1,
+    Get3C1,
+
+    Get0C2,
+    Get1C2,
+    Get2C2,
+    Get3C2,
+
+    Get0B,
+    Get1B,
+    Get2B,
+    Get3B,
+
+    // Assert that K index is not known for cache N
+    None0C1,
+    None1C1,
+    None2C1,
+    None3C1,
+
+    None0C2,
+    None1C2,
+    None2C2,
+    None3C2,
+
+    None0B,
+    None1B,
+    None2B,
+    None3B,
+};
+
+void testSubtreesForShieldedType(ShieldedType type) {
+    auto attempt = [&type](std::vector<SubtreeAction> actions) {
+        auto tree0 = RandomSubtree(0);
+        auto tree1 = RandomSubtree(1);
+        auto tree2 = RandomSubtree(2);
+        auto tree3 = RandomSubtree(3);
+
+        CCoinsViewTest base;
+        SelectParams(CBaseChainParams::REGTEST);
+        base.memorydb = new CCoinsViewDB(1 << 23, true);
+        CCoinsViewCache cache1(&base);
+        CCoinsViewCache cache2(&cache1);
+
+        for (auto action : actions) {
+            switch (action) {
+                case FlushCache1:
+                {
+                    cache1.Flush();
+                }
+                break;
```

Comment:
```suggestion
                case FlushCache1: cache1.Flush(); break;
```
and similarly for the other single-line case branches.